alws/utils/pulp_utils.py

```diff
@@ -12,10 +15,131 @@
     RpmPackage,
 )
 
+from alws.utils.modularity import IndexWrapper
+
 
 def get_uuid_from_pulp_href(pulp_href: str) -> uuid.UUID:
     return uuid.UUID(pulp_href.split("/")[-2])
 
+def get_modules_yaml_from_repo(repo_name: str):
+    base_url = "https://build.almalinux.org/pulp/content/prod/"
+    repo_url = urllib.parse.urljoin(base_url, f"{repo_name}/repodata/")
+    response = requests.get(repo_url)
+    try:
+        response.raise_for_status()
+    except Exception as exc:
+        raise(exc)
+    template_href = next(
+        (
+            line.strip()
+            for line in response.text.splitlines()
+            if line and "modules" in line
+        ),
+        None
+    )
+    if not template_href:
+        return
+    template_href = re.search(r'href="(.+-modules.yaml)"', template_href).group(1)
+    response = requests.get(urllib.parse.urljoin(repo_url, template_href))
+    response.raise_for_status()
+    return response.text
+
+# TODO: After ALBS-1012 is fixed, we can refactor this function
+# to get module packages from pulp wihtout having to grab the actual
+# modules.yaml file from the repository
+def get_rpm_module_packages_from_repository(
+    repo_id: uuid.UUID,
+    module: str,
+    pkg_names: typing.List[str] = None,
+    pkg_versions: typing.List[str] = None,
+    pkg_epochs: typing.List[str] = None,
+) -> typing.List[RpmPackage]:
+    repo_query = (
+        select(CoreRepository)
+        .where(CoreRepository.pulp_id == repo_id)
+    )
+    with get_pulp_db() as pulp_db:
+        repo = pulp_db.execute(repo_query).scalars().first()
+        repo_name = repo.name
+
+    if not repo_name:
+        return
+
+    # TODO: Getting modules.yaml files from BS production repos is not right.
+    # The problem here is that we need a way to get packages from
+    # the provided module:stream, and pulp actually doesn't have an artifact
+    # for it. This is a side effect of our current modules workflow.
+    # When building/releasing/publishing modules, we don't actually get
+    # them added into pulp, and we should do it.
+    # At this moment, we can only trust the modules that are in production
+    # repositories.
+    try:
+        repo_modules_yaml = get_modules_yaml_from_repo(repo_name)
+    except:
+        return
+
+    module_name, module_stream = module.split(':')
+    try:
+        repo_module = IndexWrapper.from_template(repo_modules_yaml).get_module(
+            module_name,
+            module_stream
+        )
+    except:
+        return
+
+    def extract_release(pkg: str):
+        regex = re.compile(
+            # name-epoch:version-
+            r'^[\w\d\-]+-\d+:[\d\.\w]+-' \
+            # release (this is the capturing group)
+            r'([\d\w\.\-]+\+(?:\w)?\d{1,5}\+(?:\w)?[\d\w]+)(?:\.[\d\w]+)' \
+            r'?(?:\.[\d\w]+)?\.' \
+            # .architecture
+            r'(?:i686|x86_64|src|noarch|aarch64|ppc64le|s390x)?$'
+        )
+        return regex.match(pkg).group(1)
```

Comment:
maybe it could be better to parse NEVRA with `hawkey`?
https://github.com/AlmaLinux/albs-web-server/blob/master/alws/utils/parsing.py#L59

Reply:
Right, thanks!